### PR TITLE
Add GPU support using NVIDIA cuML

### PR DESCRIPTION
**Pull request by bharven**
## Summary

 - Add gpu_hdbscan and gpu_umap parameters to allow users to specify whether they want to use GPU (uses NVIDIA cuML)
 - update fit and predict methods to support GPU
 - change default HDBSCAN distance metric to hamming (jaccard needs sparse input)
 - added static helper methods to fit umap / hdbscan on appropriate device

### CODE REVIEW

**Comment by bharven** [comment]
@momonga-ml right now separate parameters are used to specify whether the user wants to use gpu for umap and hdbscan - idea was that a user may want to keep single-threaded cpu umap for maximum reproducibility and gpu hdbscsan. let me know what you think, we could alternatively consolidate into one "use_gpu" param that will apply to both umap and hdbscan.

File: denseclus/DenseClus.py [hunk 1]
```diff
@@ -34,6 +34,22 @@
 import umap.umap_ as umap
 from sklearn.base import BaseEstimator, ClassifierMixin
 
+
+# CUML UMAP and HDBSCAN integration inspired by and adapted from https://github.com/ddangelov/Top2Vec/tree/master
```
Comment (smart-patrol) [inline denseclus/DenseClus.py:38]: @bharven Instead of try/except can you try something with importlib or pkgutil? For importlib it would be something like:
 ```import importlib.util

_HAVE_CUML = importlib.util.find_spec('cuml') is not None
_HAVE_CUHDBSCAN = importlib.util.find_spec('cuml.cluster.HDBSCAN') is not None
```

File: denseclus/DenseClus.py [hunk 2]
```diff
@@ -254,6 +296,32 @@ def fit(self, df: pd.DataFrame) -> None:
         logger.info("Fitting HDBSCAN...")
         self._fit_hdbscan()
 
+    @staticmethod
```
Comment (smart-patrol) [inline denseclus/DenseClus.py:299]: @bharven Can we have it so that it uses `self` and is a class method? Static makes it so that it might as well be a function.

File: denseclus/DenseClus.py [hunk 3]
```diff
@@ -332,6 +402,36 @@ def _umap_embeddings(self) -> None:
             logger.error("Invalid UMAP combine method: %s", self.umap_combine_method)
             raise ValueError("Select valid UMAP combine method")
 
+    @staticmethod
```
Comment (smart-patrol) [inline denseclus/DenseClus.py:405]: Same here

**Review by smart-patrol — CHANGES_REQUESTED** [review]
@bharven  Two small nits.

**Review by momonga-ml — APPROVED** [review]
Looks good.
